File: 03-spring-examples/apps/utils/src/main/java/com/nalepa/demo/common/monitored/customSimpleMonitoredExecutor/MonitoredExecutorService.java
package com.nalepa.demo.common.monitored.customSimpleMonitoredExecutor;

import io.micrometer.core.instrument.MeterRegistry;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.*;
import java.util.stream.Collectors;

public class MonitoredExecutorService implements ExecutorService {

    private final ExecutorService delegate;
    private final String losMessagePrefix;
    private final String name;
    private final MeterRegistry meterRegistry;

    public MonitoredExecutorService(ExecutorService delegate, String losMessagePrefix, String name, MeterRegistry meterRegistry) {
        this.delegate = delegate;
        this.losMessagePrefix = losMessagePrefix;
        this.name = name;
        this.meterRegistry = meterRegistry;
    }

    @Override
    public void shutdown() {
        delegate.shutdown();
    }

    @Override
    public List<Runnable> shutdownNow() {
        return delegate.shutdownNow();
    }

    @Override
    public boolean isShutdown() {
        return delegate.isShutdown();
    }

    @Override
    public boolean isTerminated() {
        return delegate.isTerminated();
    }

    @Override
    public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
        return delegate.awaitTermination(timeout, unit);
    }

    @Override
    public <T> Future<T> submit(Callable<T> task) {
        return delegate.submit(new MonitoredCallable<>(task, losMessagePrefix, name, meterRegistry));
    }

    @Override
    public <T> Future<T> submit(Runnable task, T result) {
        return delegate.submit(new MonitoredRunnable(task, losMessagePrefix, name, meterRegistry), result);
    }

    @Override
    public Future<?> submit(Runnable task) {
        return delegate.submit(new MonitoredRunnable(task, losMessagePrefix, name, meterRegistry));
    }

    @Override
    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
        return delegate.invokeAll(tasks.stream()
                .map(t -> new MonitoredCallable<>(t, losMessagePrefix, name, meterRegistry))
                .collect(Collectors.toList()));
    }

    @Override
    public <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException {
        return delegate.invokeAll(tasks.stream()
                .map(t -> new MonitoredCallable<>(t, losMessagePrefix, name, meterRegistry))
                .collect(Collectors.toList()), timeout, unit);
    }

    @Override
    public <T> T invokeAny(Collection<? extends Callable<T>> tasks) throws InterruptedException, ExecutionException {
        return delegate.invokeAny(tasks.stream()
                .map(t -> new MonitoredCallable<>(t, losMessagePrefix, name, meterRegistry))
                .collect(Collectors.toList()));
    }

    @Override
    public <T> T invokeAny(Collection<? extends Callable<T>> tasks, long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
        return delegate.invokeAny(tasks.stream()
                .map(t -> new MonitoredCallable<>(t, losMessagePrefix, name, meterRegistry))
                .collect(Collectors.toList()), timeout, unit);
    }

    @Override
    public void execute(Runnable command) {
        delegate.execute(new MonitoredRunnable(command, losMessagePrefix, name, meterRegistry));
    }
}

File: spring-examples/apps/utils/src/main/kotlin/com/nalepa/demo/common/monitored/customMonitoredExecutor/MonitoredExecutorService.kt
package com.nalepa.demo.common.monitored.customMonitoredExecutor

import io.micrometer.core.instrument.MeterRegistry
import java.util.concurrent.Callable
import java.util.concurrent.ExecutorService
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit


class MonitoredExecutorService(
    private val delegate: ExecutorService,
    private val losMessagePrefix: String,
    private val name: String,
    private val meterRegistry: MeterRegistry,
) : ExecutorService {
    override fun shutdown() {
        return delegate.shutdown()
    }

    override fun shutdownNow(): List<Runnable> {
        return delegate.shutdownNow()
    }

    override fun isShutdown(): Boolean {
        return delegate.isShutdown
    }

    override fun isTerminated(): Boolean {
        return delegate.isTerminated
    }

    override fun awaitTermination(timeout: Long, unit: TimeUnit): Boolean {
        return delegate.awaitTermination(timeout, unit)
    }

    override fun <T> submit(task: Callable<T>): Future<T> {
        return delegate.submit(MonitoredCallable(task, losMessagePrefix, name, meterRegistry))
    }

    override fun <T> submit(task: Runnable, result: T): Future<T> {
        return delegate.submit(MonitoredRunnable(task, losMessagePrefix, name, meterRegistry), result)
    }

    override fun submit(task: Runnable): Future<*> {
        return delegate.submit(MonitoredRunnable(task, losMessagePrefix, name, meterRegistry))
    }

    override fun <T> invokeAll(tasks: Collection<Callable<T>>): List<Future<T>> {
        return delegate.invokeAll(tasks.map { MonitoredCallable(it, losMessagePrefix, name, meterRegistry) })
    }

    override fun <T> invokeAll(
        tasks: Collection<Callable<T>>,
        timeout: Long,
        unit: TimeUnit
    ): List<Future<T>> {
        return delegate.invokeAll(
            tasks.map { MonitoredCallable(it, losMessagePrefix, name, meterRegistry) },
            timeout,
            unit
        )
    }

    override fun <T> invokeAny(tasks: Collection<Callable<T>>): T {
        return delegate.invokeAny(tasks.map { MonitoredCallable(it, losMessagePrefix, name, meterRegistry) })
    }

    override fun <T> invokeAny(
        tasks: Collection<Callable<T>>,
        timeout: Long,
        unit: TimeUnit
    ): T {
        return delegate.invokeAny(
            tasks.map { MonitoredCallable(it, losMessagePrefix, name, meterRegistry) },
            timeout,
            unit
        )
    }

    override fun execute(command: Runnable) {
        delegate.execute(MonitoredRunnable(command, losMessagePrefix, name, meterRegistry))
    }
}